akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1429,10 +1480,14 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
       _sink.pullSubstream()
     }
 
-    def cancel(): Unit = cancel(SubscriptionWithCancelException.NoMoreElementsNeeded)
+    def cancel(): Unit = {
+      cancel(SubscriptionWithCancelException.NoMoreElementsNeeded)
+      GraphStageLogic.this.completedOrFailed(this)
```

Comment:
Is that needed? Seems to delegate to the below method.

Reply:
Ups, indeed!